src/persist-txn/src/operator.rs

```diff
@@ -0,0 +1,394 @@
+                Arc::new(StringSchema::default()),
+                Diagnostics::from_purpose("txn remap"),
+            )
+            .await
+            .expect("schema shouldn't change");
+        let mut txns_cache = TxnsCache::open(T::minimum(), txns_read).await;
+
+        tracing::info!("WIP {} {} as_of {:?}", name, op_id, as_of);
+        let as_of = as_of
+            .expect("WIP do we always happen to set this?")
+            .as_option()
+            .expect("WIP")
+            .clone();
+        txns_cache.update_gt(&as_of).await;
+
+        let mut target = match txns_cache.to_data_inclusive(&data_id, as_of.clone()) {
+            Ok(translated_as_of) => {
+                tracing::info!("WIP {} {} as_of {:?} got mapped to {:?}", name, op_id, as_of, translated_as_of);
+                // WIP HACKS force the as_of to be available so the
+                // persist_source that is feeding into this doesn't block when
+                // we ask for a snapshot at that as_of
+                //
+                // WIP also regression test for this
+                let mut data_write = client.open_writer::<K, V, T, D>(
+                    data_id,
+                    data_key_schema,
+                    data_val_schema,
+                    Diagnostics::from_purpose("data advance as_of"),
+                )
+                .await
+                .expect("schema shouldn't change");
+                crate::empty_caa(|| "WIP", &mut data_write, as_of.clone()).await;
+                as_of.clone()
+            },
+            Err(_) => {
+                // WIP HACKS we seem to get reads from before the table was
+                // registered. if that happens advance to when the table was
+                // registered.
+                tracing::info!(
+                    "WIP attempting to replace bad as_of {:?} for {} {:.9}",
+                    as_of,
+                    name,
+                    data_id.to_string()
+                );
+                let mut txns_write = client
+                .open_writer::<ShardId, String, T, i64>(
+                    txns_id,
+                    Arc::new(TodoSchema::<ShardId>::default()),
+                    Arc::new(StringSchema::default()),
+                    Diagnostics::from_purpose("txn as_of"),
+                )
+                .await
+                .expect("schema shouldn't change");
+                let txns_upper = txns_write.fetch_recent_upper().await;
+                tracing::info!(
+                    "WIP attempting to replace bad as_of {:?} for {} {:.9} txns upper {:?}",
+                    as_of,
+                    name,
+                    data_id.to_string(),
+                    txns_upper
+                );
+                let txns_upper = txns_upper.as_option().expect("WIP");
+                txns_cache.update_ge(txns_upper).await;
+                let ret = txns_cache.data_since(&data_id).expect("table registered");
```

Comment:
This triggered in the Nightly Canary CI step and ended up in Sentry:

https://materializeinc.sentry.io/issues/4386718273/?project=6780145